File: web/src/queries/questions.ts
/*
 * Copyright (c) [2024] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import React from "react";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { useInstallerClient } from "~/context/installer";
import { Question } from "~/types/questions";
import { fetchQuestions, updateAnswer } from "~/api/questions";

const questionsKeys = {
  all: () => ["questions"] as const,
};

/**
 * Query to retrieve questions
 */
const questionsQuery = () => ({
  queryKey: ["questions"],
  queryFn: fetchQuestions,
});

/**
 * Hook that builds a mutation given question, allowing to answer it

 * TODO: improve/simplify it once the backend API is improved.
 */
const useQuestionsConfig = () => {
  const queryClient = useQueryClient();
  const query = {
    mutationFn: (question: Question) => updateAnswer(question),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ["questions"] }),
  };
  return useMutation(query);
};

/**
 * Hook for listening questions changes and performing proper invalidations
 */
const useQuestionsChanges = () => {
  const queryClient = useQueryClient();
  const client = useInstallerClient();

  React.useEffect(() => {
    if (!client) return;

    return client.onEvent((event) => {
      if (event.type === "QuestionsChanged") {
        queryClient.invalidateQueries({ queryKey: questionsKeys.all() });
      }
    });
  }, [client, queryClient]);

  React.useEffect(() => {
    if (!client) return;

    return client.onConnect(() => {
      queryClient.invalidateQueries({ queryKey: questionsKeys.all() });
    });
  }, [client, queryClient]);
};

/**
 * Hook for retrieving available questions
 */
const useQuestions = () => {
  const { data: questions, isPending } = useQuery(questionsQuery());
  return isPending ? [] : questions;
};

export { questionsQuery, useQuestions, useQuestionsConfig, useQuestionsChanges };

File: service/lib/agama/http/clients/questions.rb
# frozen_string_literal: true

# Copyright (c) [2025] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "agama/http/clients/base"
require "agama/question"

module Agama
  module HTTP
    module Clients
      # HTTP client to interact with the files API.
      class Questions < Base
        class CouldNotAddQuestion < StandardError; end

        # Adds a question
        #
        # @param question [Agama::Question]
        # @return [Question, nil] created question or nil if the request failed
        def add(question)
          response = post("v2/questions", question.to_api)
          response ? Question.from_api(JSON.parse(response.body)) : nil
        end

        def questions
          JSON.parse(get("v2/questions")).map { |q| Question.from_api(q) }
        end

        # Deletes the given question
        #
        # @param id [Integer] question ID
        # @return [void]
        def delete(id)
          payload = { "delete" => { "id" => id } }
          patch("v2/questions", payload)
        end

        # Waits until specified question is answered
        #
        # @param id [Integer] question ID
        # @return [void]
        def wait_answer(id)
          @logger.info "Waiting for question #{id} to be answered"

          # TODO: detect if no UI showed up to display the questions and time out?
          # for example:
          # (0..Float::INFINITY).each { |i| break if i > 100 && !question.displayed; ... }
          loop do
            found = questions.find { |q| q.id == id }
            # raise an error if the question is not found.
            return found.answer if found&.answer

            sleep(0.5)
          end
        end

        # Asks the given question and waits until the question is answered
        #
        # @example
        #   ask(question1) #=> Symbol
        #   ask(question2) { |a| a.action == :yes }  #=> Boolean
        #
        # @param question [Agama::Question]
        # @yield [Agama::Answer] Gives the answered question to the block.
        # @return [Agama::Answer] The question answer, or the result of the block in case a block
        #   is given.
        def ask(question)
          added_question = add(question)
          if added_question.nil?
            @logger.error "Could not add a question with data: {question.inspect}"
            raise CouldNotAddQuestion
          end

          answer = wait_answer(added_question.id)

          @logger.info("#{added_question.text} #{answer.inspect}")

          result = block_given? ? yield(answer) : answer
          delete(added_question.id)

          result
        end
      end
    end
  end
end
